### PR TITLE
Shows how to get a file in slices async and join the slices back again.

### PR DESCRIPTION
**Pull request by siewmoi**

### CODE REVIEW

**Comment by Zlatkovsky** [comment]
@siewmoi , can you first get @kbrandl to review?  I'm happy to take a look after as well, I just want to get into a mode where I'm an optional reviewer, not the primary one...

**Comment by kbrandl** [comment]
@Siew Moi Khor<mailto:smkhor@microsoft.com> – I will review this first thing tomorrow morning (Tuesday).

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+                if (Office.context.requirements.isSetSupported("ExcelApi", 1.7)) {
+                    sheet.getUsedRange().format.autofitColumns();
+                    sheet.getUsedRange().format.autofitRows();
+                }
+
+                sheet.activate();
+
+                await context.sync();
+                createChart();
```
Comment (Zlatkovsky) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: You shouldn't be doing nested Excel.run-s.  Just call createChart before the sync, and pass the context as a parameter

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 2]
```diff
@@ -0,0 +1,201 @@
+id: excel-82-document-get-file-in-slices-async-
+name: Get file in slices and recombine
```
Comment (kbrandl) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: Suggest that you change the name to:  Get file

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+description: Get current file in slices and recombine the slices afterwards.
```
Comment (kbrandl) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: Suggest that you change the description to:   Get the base64-encoded string that represents the current document.

(The info about getting slices, and recombining slices, is implementation details that you can mention in the snippet itself...it's not needed in the description.)

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                });
+
+            function onError(error: Office.Error): void {
+                OfficeHelpers.UI.notify('Error', error.message);
+                OfficeHelpers.Utilities.log(error);
+            }
+
+            function onSuccess(byteArray: number[]) {
+                // Now that all the file content is stored in the "data" parameter,
+                // you can do something with it, such as print, store into database, etc.
```
Comment (kbrandl) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: Suggest that you change the comment to:

// Now that all of the file content is stored in the "data" parameter,
// you can do something with it, such as print the file, store the file in a database, etc.

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                console.log("Received the full contents of the file.");
+
+                let base64string = base64js.fromByteArray(byteArray);
+                $('#file-contents').val(base64string).show();
+
+                OfficeHelpers.UI.notify("You can validate the file contents in the text box using an online tool like https://www.base64decode.org/.  To do this, copy the base64 string and paste it into the tool decoder.  Decode and open the resulting file.");
```
Comment (kbrandl) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: Suggest that you edit the text to read:

The base64-encoded string that represents the current document has been written to the text box below. To validate the string, use an online tool such as https://www.base64decode.org/ to decode the string, and then open the resulting file.

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        declare namespace base64js {
+            /** Takes a byte array and returns a base64 string
+             * Imported from https://www.npmjs.com/package/base64-js package. */
+            function fromByteArray(array: number[]): string;
+        }
+    language: typescript
+template:
+    content: |
+        <section class="ms-font-m">
+            <p>This sample shows how to get a file in slices, and join the slices back again later.</p>
```
Comment (kbrandl) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: Suggest that you change the description here to:

This sample shows how to get the base64-encoded string that represents the current document. It uses the getFileAsync() method to read the file in slices and then joins all slices back together to form the complete file.

File: samples/excel/82-document/get-file-in-slices-async.yaml (hunk 2 above)
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Done.

File: samples/excel/82-document/get-file-in-slices-async.yaml (hunk 3 above)
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done.

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+api_set:
+    ExcelApi: 1.1
+script:
+    content: |-
+        $("#setup").click(() => tryCatch(setup));
+        $("#get-file").click(() => tryCatch(getCurrentFile));
+
+
+        function getCurrentFile() {
+            const sliceSize = 4096 /*Bytes*/
```
[the remark this answers is not in the document]
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done.

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+            // Note: If you don't specify the slice size, a default will be assumed for you.  In this sample. to demonstrate how the slices work, a small size limit is specified.
```
[the remark this answers is not in the document]
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done, but omit default slice size (diff for diff platform) and mention of optional parameter.

File: samples/excel/82-document/get-file-in-slices-async.yaml (hunk 4 above)
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done.

File: samples/excel/82-document/get-file-in-slices-async.yaml (hunk 5 above)
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done.

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            function getFileContentsHelper() {
+                file.getSliceAsync(fileSlices.length, function (result) {
+                    if (result.status === Office.AsyncResultStatus.Failed) {
+                        file.closeAsync();
+                        return onError(result.error);
+                    }
+
+                    // Got one slice, store it in a temporary array.
+                    // (Or you can do something else, such as
+                    // send it to a third-party server.)
```
[the remark this answers is not in the document]
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done.

File: samples/excel/82-document/get-file-in-slices-async.yaml (hunk 6 above)
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done.

File: samples/excel/82-document/get-file-in-slices-async.yaml (hunk 1 above)
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Thanks, done.

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                }
+
+                sheet.activate();
+                createChart(context);
+
+                await context.sync();                
+            });
+        } 
+
+        async function createChart(context) {
```
Comment (Zlatkovsky) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: Please add "context: Excel.RequestContext" annotation for the parameter, for IntelliSense

File: samples/excel/82-document/get-file-in-slices-async.yaml [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                    ["Mirrors", 225, 600, 923, 544],
+                    ["Spokes", 6005, 7634, 4589, 8765]
+                ]);
+
+                if (Office.context.requirements.isSetSupported("ExcelApi", 1.7)) {
+                    sheet.getUsedRange().format.autofitColumns();
+                    sheet.getUsedRange().format.autofitRows();
+                }
+
+                sheet.activate();
```
Comment (Zlatkovsky) [inline samples/excel/82-document/get-file-in-slices-async.yaml]: I would move sheet.activate down by one (to only activate after chart creation -- I.e., as the very final call before the sync

Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Incorped, thanks!

File: samples/excel/82-document/get-file-in-slices-async.yaml (hunk 10 above)
Comment (siewmoi) [reply samples/excel/82-document/get-file-in-slices-async.yaml]: Incorped, thanks!

**Pull request merged**